src/object/srv_obj.c

```diff
@@ -1803,15 +1803,23 @@ enum process_epoch_rc {
 static int
 process_epoch(uint64_t *epoch, uint64_t *epoch_first, uint32_t *flags)
 {
-	if (*epoch == 0 || *epoch == DAOS_EPOCH_MAX)
+	if (*epoch == 0 || *epoch == DAOS_EPOCH_MAX) {
 		/*
 		 * *epoch is not a chosen TX epoch. Choose the current HLC
 		 * reading as the TX epoch.
 		 */
 		*epoch = crt_hlc_get();
-	else
+	} else {
+		/*
+		 * If the TX is started before current server (re-)start,
+		 * then needs to be restarted with newer epoch.
+		 */
+		if (*epoch < dss_get_start_epoch())
+			return -DER_TX_RESTART;
```

Comment:
Do we really want this to apply to reads?

Reply:
In fact, we can only check for CPD RPC. The check here (for non-stand-along read) may detect the potential conflict some early to avoid unnecessary TX operations. But I am not sure which one will cause more overhead, depends on the server restart frequency.

Reply:
Agreed.